x/gamm/types/pool.go

```diff
@@ -19,22 +19,31 @@ type PoolI interface {
 	// Todo: may have to generalize this later
 	// Also todo: come up with better name here
 	GetSwapFee(ctx sdk.Context) sdk.Dec
+	GetExitFee(ctx sdk.Context) sdk.Dec
 	// Todo: These may not be needed
 	NumAssets() int
 	IsActive(curBlockTime time.Time) bool
 	// Returns the coins in the pool owned by all LP shareholders
 	GetTotalLpBalances(ctx sdk.Context) sdk.Coins
+	// TODO: Add ctx here
 	GetTotalShares() sdk.Int
 
 	CalcOutAmtGivenIn(ctx sdk.Context, tokenIn sdk.Coins, tokenOutDenom string, swapFee sdk.Dec) (tokenOut sdk.DecCoin, err error)
 	CalcInAmtGivenOut(ctx sdk.Context, tokenOut sdk.Coins, tokenInDenom string, swapFee sdk.Dec) (tokenIn sdk.DecCoin, err error)
 
+	// TODO: Ensure this can only be called via gamm
+	// TODO: Think through the API guarantees this is providing in conjunction with the caller being
+	// expected to Set the pool into state as well.
+	ApplySwap(ctx sdk.Context, tokenIn sdk.Coins, tokenOut sdk.Coins) error
```

Comment:
Great interface! Maybe add `shareIn, shareOut sdk.Int` to capture shareswap case?

Reply:
Interesting idea. At the moment JoinPool / ExitPool be mutative. (This is because I made iterative logic in the join pool code, which would be harder if it was non mutative)

wdyt about those being mutative, vs an "apply liquidity change" method?